.github/workflows/release.yaml

```diff
@@ -0,0 +1,152 @@
+name: Release
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, labeled, unlabeled]
+    branches: [main]
+    paths:
+      - "*/Cargo.toml" # only run if Cargo.toml changes
+  pull_request_target:
+    types: [closed]
```

Comment:
Why do we want to run when a PR is closed?

Reply:
I think we more likely want:
```yaml
on:
  pull_request:
  push:
    branches:
      - main
```

So we run on:
* Any PR (no need for types, we want them all I think)
* Any push to main (== PR merged)

Could make it
```yaml
on:
  pull_request:
    branches:
      - main
  push:
    branches:
      - main
```
which would only run on PRs directly targeting main (so not stacked PRs).

Reply:
For your questions:

> Why do we want to run when a PR is closed?

Normally we can use `on: push`, but it will set the actions context to a branch instead of a PR, which would make it hard (but not impossible) for me to look up PR information, such as labelling. In my design, I check for whether the label `release` is attached to the PR such that we can control whether we are intentionally trying to release something.

I have done `look up PR information from the main branch after merge` in the past, but the solution is very janky. I was under different constraints back then, when that was the only viable solution. I would not suggest we do that here.

`pull_request_target` is a nice alternative because it has the context of a PR but runs on the base branch (i.e. the SHA of the commit would be in the main branch, instead of the last commit in a PR). And I have a guard to only disable dry-run when the PR is merged.

---

> Any PR (no need for types, we want them all I think)

The [GitHub Actions docs](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request) says:

```
By default, a workflow only runs when a pull_request event's activity type is opened, synchronize, or reopened
```

Which means I won't get labeled/unlabeled events by default, so I have to specify them manually.

---

> which would only run on PRs directly targeting main

With my current `pull_request_target` config, the action would still run when someone closes the PR without merging. But because of my dry-run check, it will run in the dry-run mode which is harmless.

Reply:
> , which would make it hard (but not impossible) for me to look up PR information, such as labelling

https://github.com/jwalton/gh-find-current-pr could be used? We use successfully in `.github/workflows/ci.yaml`.

> pull_request_target is a nice alternative because it has the context of a PR but runs on the base branch (i.e. the SHA of the commit would be in the main branch, instead of the last commit in a PR). And I have a guard to only disable dry-run when the PR is merged.

So on a PR we run on main and not on the PR? That seems odd on first glance.

I think maybe adding some "How to do a release" documentation might help clarify things and also be useful for anyone who comes to make a release in the future.


> Which means I won't get labeled/unlabeled events by default, so I have to specify them manually.

Thanks, I hadn't realized about the release label stuff, BTW In the commit message you said "release tag" (if I'd thought about it I'd have realized that was odd and you probably meant label, but I didn't...)

Reply:
> https://github.com/jwalton/gh-find-current-pr could be used?

I think I have implemented `jwalton/gh-find-current-pr` on my own before, that is exactly the janky solution I was talking about. Although I believe `pull_request_target` is the right way of getting PR context, `on: push: branches: [main]`is more intuitive which would prevent us from making mistakes in the future. Let me make changes to that.

> I think maybe adding some "How to do a release" documentation

Adding documentation on how to do a release is definitely on my TODO list. And I will make sure to add PR descriptions I make it ready to review. Btw thanks for the early review!

> In the commit message you said "release tag" 

Let me update the commit messages. I must have mixed up with the wordings.